README.md
# Merritt::Manifest

[![Build Status](https://travis-ci.org/dmolesUC3/merritt-manifest.svg)](https://travis-ci.org/dmolesUC3/merritt-manifest)
[![Inline docs](http://inch-ci.org/github/dmolesUC3/merritt-manifest.svg)](http://inch-ci.org/github/dmolesUC3/merritt-manifest)

A library for creating [Merritt](https://merritt.cdlib.org/) [text-based manifests](https://merritt.cdlib.org/help/manifest_guide#text_tips) (a specialization of the [Checkm](https://wiki.ucop.edu/display/Curation/Checkm) text-based manifest format).

See the [Merritt ingest service docs](https://confluence.ucop.edu/download/attachments/16744573/Merritt-ingest-service-latest.pdf)
for more information.

## Object support

Single-object manifests are supported via the `Merritt::Manifest::Object` class:

```ruby
obj_manifest = Merritt::Manifest::Object.new(
  files: [
    {
      file_url: 'http://merritt.cdlib.org/samples/4blocks.jpg',
      hash_algorithm: 'md5',
      hash_value: '0b21c6d48e815dd537d42dc1cfac0111',
      file_name: '4blocks.jpg'
    },
    {
      file_url: 'http://merritt.cdlib.org/samples/4blocks.txt',
      hash_algorithm: 'md5',
      hash_value: 'ed04a855f89f31f8dc8e9bb946f5f159',
      file_name: '4blocks.txt'
    }
  ].map { |h| OpenStruct.new(h) }
)

puts obj_manifest.write_to_string
```

```text
#%checkm_0.7
#%profile | http://uc3.cdlib.org/registry/ingest/manifest/mrt-ingest-manifest
#%prefix | mrt: | http://merritt.cdlib.org/terms#
#%prefix | nfo: | http://www.semanticdesktop.org/ontologies/2007/03/22/nfo#
#%fields | nfo:fileUrl | nfo:hashAlgorithm | nfo:hashValue | nfo:fileSize | nfo:fileLastModified | nfo:fileName | mrt:mimeType
http://merritt.cdlib.org/samples/4blocks.jpg | md5 | 0b21c6d48e815dd537d42dc1cfac0111 |  |  | 4blocks.jpg
http://merritt.cdlib.org/samples/4blocks.txt | md5 | ed04a855f89f31f8dc8e9bb946f5f159 |  |  | 4blocks.txt
#%eof
```

## Batch support

Batch manifests are not yet supported.

## DataONE Manifest support

DataONE manifests are supported via the `Merritt::Manifest::DataONE` class:

```ruby
dataone_manifest = Merritt::Manifest::DataONE.new(
  files: {
    'Laney_300394_Exempt_Determination_Letter.pdf' => 'application/pdf',
    'Laney_IRBProposal.docx' => 'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
    'SensorSurvey_Printout.pdf' => 'application/pdf',
    'clean_survey_data_no_ids.csv' => 'text/csv',
    'research_coords.csv' => 'text/csv',
    'survey/Q10/research_sites.R' => 'text/plain',
    'survey/Q11-23/sensors_platforms.R' => 'text/plain',
    'survey/Q24/limitsToExpansion.R' => 'text/plain',
    'survey/Q25-32/data_metadata_management.R' => 'text/plain',
    'survey/Q3-9/respondent_info.R' => 'text/plain',
    'survey/Q33-37/networking.R' => 'text/plain',
    'survey/Q38-42/publications.R' => 'text/plain',
    'survey_data_prep.R' => 'text/plain'
  }.map { |name, type| OpenStruct.new(name: name, type: type) }
)

puts dataone_manifest.write_to_string
```

```text
#%dataonem_0.1
#%profile | http://uc3.cdlib.org/registry/ingest/manifest/mrt-dataone-manifest
#%prefix | dom: | http://uc3.cdlib.org/ontology/dataonem
#%prefix | mrt: | http://uc3.cdlib.org/ontology/mom
#%fields | dom:scienceMetadataFile | dom:scienceMetadataFormat | dom:scienceDataFile | mrt:mimeType
mrt-datacite.xml | http://datacite.org/schema/kernel-3.1 | Laney_300394_Exempt_Determination_Letter.pdf | application/pdf
mrt-oaidc.xml | http://dublincore.org/schemas/xmls/qdc/2008/02/11/qualifieddc.xsd | Laney_300394_Exempt_Determination_Letter.pdf | application/pdf
mrt-datacite.xml | http://datacite.org/schema/kernel-3.1 | Laney_IRBProposal.docx | application/vnd.openxmlformats-officedocument.wordprocessingml.document
mrt-oaidc.xml | http://dublincore.org/schemas/xmls/qdc/2008/02/11/qualifieddc.xsd | Laney_IRBProposal.docx | application/vnd.openxmlformats-officedocument.wordprocessingml.document
mrt-datacite.xml | http://datacite.org/schema/kernel-3.1 | SensorSurvey_Printout.pdf | application/pdf
mrt-oaidc.xml | http://dublincore.org/schemas/xmls/qdc/2008/02/11/qualifieddc.xsd | SensorSurvey_Printout.pdf | application/pdf
mrt-datacite.xml | http://datacite.org/schema/kernel-3.1 | clean_survey_data_no_ids.csv | text/csv
mrt-oaidc.xml | http://dublincore.org/schemas/xmls/qdc/2008/02/11/qualifieddc.xsd | clean_survey_data_no_ids.csv | text/csv
mrt-datacite.xml | http://datacite.org/schema/kernel-3.1 | research_coords.csv | text/csv
mrt-oaidc.xml | http://dublincore.org/schemas/xmls/qdc/2008/02/11/qualifieddc.xsd | research_coords.csv | text/csv
mrt-datacite.xml | http://datacite.org/schema/kernel-3.1 | survey/Q10/research_sites.R | text/plain
mrt-oaidc.xml | http://dublincore.org/schemas/xmls/qdc/2008/02/11/qualifieddc.xsd | survey/Q10/research_sites.R | text/plain
mrt-datacite.xml | http://datacite.org/schema/kernel-3.1 | survey/Q11-23/sensors_platforms.R | text/plain
mrt-oaidc.xml | http://dublincore.org/schemas/xmls/qdc/2008/02/11/qualifieddc.xsd | survey/Q11-23/sensors_platforms.R | text/plain
mrt-datacite.xml | http://datacite.org/schema/kernel-3.1 | survey/Q24/limitsToExpansion.R | text/plain
mrt-oaidc.xml | http://dublincore.org/schemas/xmls/qdc/2008/02/11/qualifieddc.xsd | survey/Q24/limitsToExpansion.R | text/plain
mrt-datacite.xml | http://datacite.org/schema/kernel-3.1 | survey/Q25-32/data_metadata_management.R | text/plain
mrt-oaidc.xml | http://dublincore.org/schemas/xmls/qdc/2008/02/11/qualifieddc.xsd | survey/Q25-32/data_metadata_management.R | text/plain
mrt-datacite.xml | http://datacite.org/schema/kernel-3.1 | survey/Q3-9/respondent_info.R | text/plain
mrt-oaidc.xml | http://dublincore.org/schemas/xmls/qdc/2008/02/11/qualifieddc.xsd | survey/Q3-9/respondent_info.R | text/plain
mrt-datacite.xml | http://datacite.org/schema/kernel-3.1 | survey/Q33-37/networking.R | text/plain
mrt-oaidc.xml | http://dublincore.org/schemas/xmls/qdc/2008/02/11/qualifieddc.xsd | survey/Q33-37/networking.R | text/plain
mrt-datacite.xml | http://datacite.org/schema/kernel-3.1 | survey/Q38-42/publications.R | text/plain
mrt-oaidc.xml | http://dublincore.org/schemas/xmls/qdc/2008/02/11/qualifieddc.xsd | survey/Q38-42/publications.R | text/plain
mrt-datacite.xml | http://datacite.org/schema/kernel-3.1 | survey_data_prep.R | text/plain
mrt-oaidc.xml | http://dublincore.org/schemas/xmls/qdc/2008/02/11/qualifieddc.xsd | survey_data_prep.R | text/plain
#%eof
```


CHANGES.md
# 0.1.2 (22 April 2017)

- Fixed issue where unknown MIME types would not be written to the manifest (#1)

# 0.1.1 (18 April 2017)

- `Merritt::Manifest::Fields::Object::MIME_TYPE` now correctly parses strings as
  `MIME::Type` objects.
- **BREAKING CHANGE:** `Merritt::Manifest::DataONE` now uses the standard `file_name`
  and `mime_type` fields rather than `name` and `type`.

# 0.1.0 (18 April 2017)

- Initial release.


spec/spec_helper.rb
# ------------------------------------------------------------
# SimpleCov setup

if ENV['COVERAGE']
  require 'simplecov'
  require 'simplecov-console'

  SimpleCov.minimum_coverage 100
  SimpleCov.start do
    add_filter '/spec/'
    SimpleCov.formatter = SimpleCov::Formatter::MultiFormatter[
      SimpleCov::Formatter::HTMLFormatter,
      SimpleCov::Formatter::Console,
    ]
  end
end

# ------------------------------------------------------------
# Rspec configuration

RSpec.configure do |config|
  config.raise_errors_for_deprecations!
  config.mock_with :rspec
end

# ------------------------------------------------------------
# Merritt::Manifest

require 'merritt'


spec/unit/merritt/manifest_spec.rb
require 'spec_helper'

module Merritt
  describe Manifest do

    attr_reader :dataone_profile

    before(:each) do
      @dataone_profile = 'http://uc3.cdlib.org/registry/ingest/manifest/mrt-dataone-manifest'
    end

    describe :new do
      it 'creates a new manifest' do
        manifest = Manifest.new(profile: dataone_profile)
        expect(manifest).to be_a(Manifest)
      end

      describe :conformance do
        it 'defaults to CheckM 0.7' do
          manifest = Manifest.new(profile: dataone_profile)
          expect(manifest.conformance).to eq('checkm_0.7')
        end
      end

      describe :profile do
        it 'accepts a string' do
          profile_str = dataone_profile
          manifest = Manifest.new(profile: profile_str)
          expect(manifest.profile).to eq(URI.parse(profile_str))
        end

        it 'accepts a URI' do
          profile = URI.parse(dataone_profile)
          manifest = Manifest.new(profile: profile)
          expect(manifest.profile).to eq(profile)
        end

        it 'requires a profile' do
          expect { Manifest.new }.to raise_error(ArgumentError)
        end

        it 'requires profile to be a URL' do
          expect { Manifest.new(profile: 'elvis') }.to raise_error(ArgumentError)
        end

        it 'requires a Merritt manifest profile' do
          profile = 'http://example.org/my-profile'
          expect { Manifest.new(profile: profile) }.to raise_error(ArgumentError)
        end
      end

      describe :prefixes do
        it 'defaults to empty' do
          manifest = Manifest.new(profile: dataone_profile)
          prefixes = manifest.prefixes
          expect(prefixes).to be_a(Hash)
          expect(prefixes).to be_empty
        end

        it 'normalizes keys to symbols and values to URIs' do
          in_prefixes = {
            :mrt => URI('http://merritt.cdlib.org/terms#'),
            'dom' => 'http://uc3.cdlib.org/ontology/dataonem',
            :nfo => 'http://www.semanticdesktop.org/ontologies/2007/03/22/nfo#'
          }
          manifest = Manifest.new(profile: dataone_profile, prefixes: in_prefixes)
          out_prefixes = manifest.prefixes
          expect(out_prefixes.size).to eq(in_prefixes.size)
          expect(out_prefixes[:mrt]).to eq(in_prefixes[:mrt])
          expect(out_prefixes[:dom]).to eq(URI(in_prefixes['dom']))
          expect(out_prefixes[:nfo]).to eq(URI(in_prefixes[:nfo]))
        end
      end

      describe :fields do
        it 'defaults to empty' do
          manifest = Manifest.new(profile: dataone_profile)
          fields = manifest.fields
          expect(fields).to be_an(Array)
          expect(fields).to be_empty
        end

        it 'accepts a list of fields' do
          in_fields = %w[dom:scienceMetadataFile dom:scienceMetadataFormat dom:scienceDataFile mrt:mimeType]
          manifest = Manifest.new(
            profile: dataone_profile,
            fields: in_fields,
            prefixes: {
              mrt: 'http://merritt.cdlib.org/terms#',
              dom: 'http://uc3.cdlib.org/ontology/dataonem'
            }
          )
          out_fields = manifest.fields
          expect(out_fields.size).to eq(in_fields.size)
          expect(out_fields).to eq(in_fields)
          expect(out_fields).not_to be(in_fields)
        end
      end

      describe :entries do
        it 'accepts a list of entries' do

          in_entries = [
            {
              'dom:scienceMetadataFile' => 'mrt-datacite.xml',
              'dom:scienceMetadataFormat' => 'http://datacite.org/schema/kernel-3.1',
              'dom:scienceDataFile' => 'Laney_300394_Exempt_Determination_Letter.pdf',
              'mrt:mimeType' => 'application/pdf'
            },
            {
              'dom:scienceMetadataFile' => 'mrt-oaidc.xml',
              'dom:scienceMetadataFormat' => 'http://dublincore.org/schemas/xmls/qdc/2008/02/11/qualifieddc.xsd',
              'dom:scienceDataFile' => 'Laney_300394_Exempt_Determination_Letter.pdf',
              'mrt:mimeType' => 'application/pdf'
            }
          ]

          manifest = Manifest.new(
            profile: dataone_profile,
            fields: %w[dom:scienceMetadataFile dom:scienceMetadataFormat dom:scienceDataFile mrt:mimeType],
            prefixes: {
              mrt: 'http://merritt.cdlib.org/terms#',
              dom: 'http://uc3.cdlib.org/ontology/dataonem'
            },
            entries: in_entries
          )

          out_entries = manifest.entries
          expect(out_entries).to eq(in_entries)
        end
      end
    end
  end
end


spec/unit/merritt/manifest/fields_spec.rb
require 'spec_helper'
require 'ostruct'

module Merritt
  describe Manifest::Fields::Object do
    describe :value_from do
      describe :FILE_URL do
        it 'returns a URI even if given a String' do
          url_str = 'http://example.org/example.txt'
          obj = OpenStruct.new(file_url: url_str)
          expected = URI(url_str)
          actual = Manifest::Fields::Object::FILE_URL.value_from(obj)
          expect(actual).to eq(expected)
        end

        it 'fails if not present' do
          obj = OpenStruct.new
          expect { Manifest::Fields::Object::FILE_URL.value_from(obj) }.to raise_error(ArgumentError)
        end
      end

      describe :FILE_SIZE do
        it 'returns nil for nil' do
          obj = OpenStruct.new
          actual = Manifest::Fields::Object::FILE_SIZE.value_from(obj)
          expect(actual).to be_nil
        end
        it 'returns an integer even if given a String' do
          obj = OpenStruct.new(file_size: '1234')
          expected = 1234
          actual = Manifest::Fields::Object::FILE_SIZE.value_from(obj)
          expect(actual).to eq(expected)
        end
      end

      describe :FILE_NAME do
        it 'prefers the filename to the URI' do
          url = URI('http://example.org/example.txt')
          name = 'EXAMPLE.TXT'
          obj = OpenStruct.new(file_url: url, file_name: name)
          actual = Manifest::Fields::Object::FILE_NAME.value_from(obj)
          expect(actual).to eq(name)
        end

        it 'extracts the filename from the URI if not present' do
          url = URI('http://example.org/example.txt')
          obj = OpenStruct.new(file_url: url)
          expected = 'example.txt'
          actual = Manifest::Fields::Object::FILE_NAME.value_from(obj)
          expect(actual).to eq(expected)
        end
      end

      describe :MIME_TYPE do
        it 'extracts the mime type' do
          mime_type = 'text/html'
          obj = OpenStruct.new(mime_type: mime_type)
          mt_val = ::MIME::Types[mime_type].first
          expect(mt_val).not_to be_nil # just to be sure
          actual = Manifest::Fields::Object::MIME_TYPE.value_from(obj)
          expect(actual).to eq(mt_val.to_s)
        end

        it 'accepts unknown or garbage types' do
          mime_type = 'I am not a mime type, but whatever'
          obj = OpenStruct.new(mime_type: mime_type)
          mt_val = ::MIME::Types[mime_type].first
          expect(mt_val).to be_nil # just to be sure
          actual = Manifest::Fields::Object::MIME_TYPE.value_from(obj)
          expect(actual).to eq(mime_type)
        end
      end
    end
  end
end


lib/merritt/manifest.rb
Dir.glob(File.expand_path('../manifest/*.rb', __FILE__)).sort.each(&method(:require))

module Merritt
  # A Merritt manifest file
  class Manifest

    # Base for all recognized profile URIs
    PROFILE_BASE_URI = 'http://uc3.cdlib.org/registry/ingest/manifest/'.freeze

    # Checkm 0.7 conformance level
    CHECKM_0_7 = 'checkm_0.7'.freeze

    # @return [String] the conformance level
    attr_reader :conformance

    # @return [URI] the profile URI
    attr_reader :profile

    # @return [Hash{Symbol => URI}] a map from namespace prefixes to their URIs
    attr_reader :prefixes

    # @return [Array<String>] the field names, in the form prefix:fieldname
    attr_reader :fields

    # @return [Array<Hash<String, Object>>] the entries
    attr_reader :entries

    # Creates a new manifest. Note that the prefix, field, and entry arrays are
    # copied on initialization, as are the individual entry hashes.
    #
    # @param conformance [String] the conformance level. Defaults to {CHECKM_0_7}.
    # @param profile [URI, String] the profile URI. Must begin with
    # @param prefixes [Hash{String,Symbol => URI, String}] a map from namespace prefixes to their URIs
    # @param fields Array<String> a list of field names, in the form prefix:fieldname
    # @param entries [Array<Hash<String, Object><] A list of entries, each of which is a hash keyed by
    #   a prefixed fieldname defined in `fields`. Nil values are allowed.
    # @raise [ArgumentError] if `profile` does not begin with {PROFILE_BASE_URI}
    # @raise [ArgumentError] if `fields` cannot be parsed as prefix:fieldname, or if one or more prefixes
    #   is not mapped to a URI in `prefixes`
    # @raise [URI::InvalidURIError] if `profile` cannot be parsed as a URI
    def initialize(conformance: CHECKM_0_7, profile:, prefixes: {}, fields: [], entries: [])
      @conformance = conformance
      @profile = normalize_profile_uri(profile).freeze
      @prefixes = normalize_prefixes(prefixes).freeze
      @fields = validate_fields(fields).freeze
      @entries = normalize_entries(entries).freeze
    end

    # Writes this manifest to the specified IO
    # @param io [IO] the IO to write to
    def write_to(io)
      write_sc(io, conformance)
      write_sc(io, 'profile', profile)
      prefixes.each { |prefix, url| write_sc(io, 'prefix', "#{prefix}:", url) }
      write_sc(io, 'fields', *fields)
      entries.each { |entry| io.puts(entry_line(entry)) }
      write_sc(io, 'eof')
    end

    # Writes this manifest as a string
    # @return [String] the manifest file contents as a string
    def write_to_string
      io = StringIO.new
      write_to(io)
      io.string
    end

    private

    # checkm column separator
    COLSEP = ' | '.freeze

    def entry_line(entry)
      fields.map { |f| entry[f] }.join(COLSEP).sub(/[| ]+\z/, '')
    end

    # writes a checkm "structured comment"
    # @param io [IO] the IO to write to
    # @param comment [String] the comment
    # @param columns [nil, Array<String>] columns to follow the initial comment
    def write_sc(io, comment, *columns)
      io << '#%' << comment
      io << COLSEP << columns.join(COLSEP) unless columns.empty?
      io << "\n"
    end

    def normalize_entries(entries)
      entries.each_with_index.map do |entry, i|
        raise ArgumentError, "Nil entry at index #{i}" unless entry
        normalize_entry(entry)
      end
    end

    def normalize_entry(entry)
      normalized = {}
      fields.each do |f|
        next unless (value = entry[f])
        normalized[f] = value
      end
      raise ArgumentError, "No fields found in entry #{entry}" if normalized.empty?
      normalized
    end

    def validate_fields(fields)
      fields.map { |f| validate_field(f) }
    end

    def validate_field(field)
      prefix, fieldname = field.split(':')
      raise ArgumentError "Unknown prefix in field '#{field}': #{prefix}" unless prefixes.key?(prefix.to_sym)
      raise ArgumentError "Field '#{field}' cannot be parsed as prefix:fieldname" unless fieldname
      field
    end

    def normalize_prefixes(prefixes)
      return {} unless prefixes
      prefixes.map { |k, v| [k.to_sym, Util.to_uri(v)] }.to_h
    end

    def normalize_profile_uri(profile)
      profile_uri = Util.to_uri(profile)
      raise ArgumentError, "Invalid profile: #{profile || 'nil'}" unless profile_uri &&
        profile_uri.to_s.start_with?(PROFILE_BASE_URI)
      profile_uri.clone # defensive copy
    end

  end
end


lib/merritt/manifest/file.rb
require 'merritt/manifest/fields'

module Merritt
  class Manifest
    # A marker interface for file-like objects. Each field
    # may or may not be relevant to a given manifest format.
    module File
      # @return [nil, URI] the URL at which to retrieve the file
      attr_reader :file_url

      # @return [nil, String] the hash algorithm used to hash the file
      attr_reader :hash_algorithm

      # @return [nil, String] the hash value
      attr_reader :hash_value

      # @return [nil, Integer] the file size in bytes.
      attr_reader :file_size

      # @return [nil, DateTime] date and time the file was last modified.
      #   Note that according to the [Merritt Ingest Service docs](https://confluence.ucop.edu/download/attachments/16744573/Merritt-ingest-service-latest.pdf),
      #   “modification time field SHOULD NOT be specified, and will be ignored if provided.”
      attr_reader :file_last_modified

      # @return [nil, String] the file name. If not present, will be inferred from
      #   {#file_url}.
      attr_reader :file_name

      # @return [nil, MIME::Type] the mime type
      attr_reader :mime_type
    end
  end
end


lib/merritt/manifest/object.rb
module Merritt
  class Manifest
    # A specialization of {Manifest} for Merritt object submissions.
    class Object < Manifest

      # Creates a new {Manifest::Object}
      # @param files [Array<Manifest::File>] an array of data files to be converted to entries.
      #   (Note that these not be actual {Manifest::File} objects so long as they respond to,
      #   at minimum, `#file_url`. The other {Manifest::File} fields are optional.)
      def initialize(files:)
        super(
          profile: 'http://uc3.cdlib.org/registry/ingest/manifest/mrt-ingest-manifest',
          prefixes: Merritt::Manifest::Fields::Object.prefixes,
          fields: Merritt::Manifest::Fields::Object.fields,
          entries: to_entries(files)
        )
      end

      private

      def to_entries(files)
        files.map do |file|
          Merritt::Manifest::Fields::Object.map do |field|
            field_name = field.field_name
            field_value = field.value_from(file)
            [field_name, field_value]
          end.to_h
        end
      end
    end
  end
end


lib/merritt/manifest/data_one.rb
require 'merritt/manifest/fields'

module Merritt
  class Manifest
    # A specialization of {Manifest} for DataONE.
    class DataONE < Manifest

      # Creates a new {Manifest::DataONE}
      # @param files [Array<Manifest::File>] an array of data files to be converted to entries.
      #   (Note that these not be actual {Manifest::File} objects so long as they respond to
      #   `#file_name` and `#mime_type`)
      def initialize(files:)
        super(
          conformance: 'dataonem_0.1',
          profile: 'http://uc3.cdlib.org/registry/ingest/manifest/mrt-dataone-manifest',
          prefixes: {
            dom: 'http://uc3.cdlib.org/ontology/dataonem',
            mrt: 'http://uc3.cdlib.org/ontology/mom'
          },
          fields: [METADATA_FILE, METADATA_FORMAT, DATA_FILE, MIME_TYPE],
          entries: to_entries(files)
        )
      end

      private

      METADATA_FILE = 'dom:scienceMetadataFile'.freeze
      METADATA_FORMAT = 'dom:scienceMetadataFormat'.freeze
      DATA_FILE = 'dom:scienceDataFile'.freeze
      MIME_TYPE = 'mrt:mimeType'.freeze

      METADATA_FILES = {
        'mrt-datacite.xml' => 'http://datacite.org/schema/kernel-3.1',
        'mrt-oaidc.xml' => 'http://dublincore.org/schemas/xmls/qdc/2008/02/11/qualifieddc.xsd'
      }.freeze

      def to_entries(files)
        rows = files.product(METADATA_FILES.to_a).map(&:flatten)
        rows.map do |file, md_name, md_url|
          {
            METADATA_FILE => md_name,
            METADATA_FORMAT => md_url,
            DATA_FILE => file.file_name,
            MIME_TYPE => Merritt::Manifest::Fields::Object::MIME_TYPE.value_from(file)
          }
        end
      end
    end
  end
end


lib/merritt/manifest/fields.rb
require 'typesafe_enum'
require 'mime-types'
require 'uri'

module Merritt
  class Manifest

    # Mixin for field enumerations
    module Field
      # Gets the reader method for this field
      def reader
        @reader ||= key.downcase
      end

      # Reads the value of this field from the specified object, if available
      def value_from(obj)
        return unless obj.respond_to?(reader)
        obj.send(reader)
      end

      # @return [String] the fieldname
      def field_name
        value
      end

      # The prefix for this field
      def prefix
        @prefix ||= begin
          prefix, name = value.split(':')
          prefix if name # if we didn't find a name, then there's no ':' and hence no prefix
        end
      end

      # Implemented to inject {ClassMethods} into field enumerations that include this module
      def self.included(base)
        base.extend(ClassMethods)
      end

      # Class methods for field enumerations
      module ClassMethods
        # Gets all prefixes in these fields, and their URLs
        # @return [Hash<Symbol, String>] all prefixes used by these fields, as a map from symbol to URL string
        def prefixes
          @prefixes ||= begin
            all_prefixes = to_a.map(&:prefix).uniq.sort.freeze
            all_prefixes.map { |p| [p, url_for(p)] }.to_h.freeze
          end
        end

        # A list of all fields
        # @return [String] a list of all fields, as (prefix-qualified) names
        def fields
          @fields ||= to_a.map(&:value).freeze
        end

        # Gets the string for the specified prefix
        # @param prefix [String, Symbol] the prefix
        # @return [String] the URL string for the prefix
        def url_for(prefix)
          # noinspection RubyCaseWithoutElseBlockInspection
          case prefix.to_sym
          when :mrt
            'http://merritt.cdlib.org/terms#'
          when :nfo
            'http://www.semanticdesktop.org/ontologies/2007/03/22/nfo#'
          end
        end
      end
    end

    # Holder module for field enumerations
    module Fields
      # Enumeration of fields for object manifests
      class Object < TypesafeEnum::Base
        include Field
        # @!parse extend Merritt::Manifest::Field::ClassMethods

        # Field for `nfo:fileUrl`. Parses string URLs as URI objects.
        new(:FILE_URL, 'nfo:fileUrl') do
          def value_from(obj)
            value = super(obj)
            raise ArgumentError, "No :#{reader} method provided for #{obj}" unless value
            Merritt::Util.to_uri(value)
          end
        end

        # field for `nfo:hashAlgorithm`
        new :HASH_ALGORITHM, 'nfo:hashAlgorithm'

        # field for `nfo:hashValue`
        new :HASH_VALUE, 'nfo:hashValue'

        # field for `nfo:fileSize`. Parses string values as integers.
        new :FILE_SIZE, 'nfo:fileSize' do
          def value_from(obj)
            value = super(obj)
            return unless value
            value.to_i
          end
        end

        # field for `nfo:fileLastModified`
        new :FILE_LAST_MODIFIED, 'nfo:fileLastModified'

        # field for `nfo:fileName`. If no file name is provided, parses it from {FILE_URL}.
        new :FILE_NAME, 'nfo:fileName' do
          def value_from(obj)
            value = super(obj)
            return value if value
            file_url = FILE_URL.value_from(obj)
            URI(file_url).path.split('/').last
          end
        end

        # field for `mrt:mimeType`
        new :MIME_TYPE, 'mrt:mimeType' do
          def value_from(obj)
            value = super(obj)
            return value if value.nil? || value.is_a?(::MIME::Type)
            ::MIME::Types[value.to_s].first || value.to_s
          end
        end
      end
    end
  end
end
